Todo:
- [x] setup git repo
- [x] create python script
- [x] function to read xfma entries files 
	- [x] return the swap to be made
- [x] function to read genomes 
- [x] function to make swap 
	- [x] takes positions
	- [x] makes swap for both genomes as acceptor and donor
- [ ] deal with strand
- [ ] output to fasta
Xmfa file format
>The .alignment file contains the complete genome alignment generated by Mauve in the eXtended Multi-FastA (XMFA) file format. This standard file format is also used by other genome alignment systems that align sequences with rearrangements. The XMFA file format supports the storage of several collinear sub-alignments, each separated with an = sign, that constitute a single genome alignment. Each sub-alignment consists of one FastA format sequence entry per genome where the entry’s defline gives the strand (orientation) and location in the genome of the sequence in the alignment.

From: https://darlinglab.org/mauve/user-guide/files.html

The file looks like this:
```
>seq_num:start1-end1 ± comments (sequence name, etc.)
AC-TG-NAC--TG
AC-TG-NACTGTG
...

> seq_num:startN-endN ± comments (sequence name, etc.)
AC-TG-NAC--TG
AC-TG-NACTGTG
...
= comments, and optional field-value pairs, i.e. score=12345
```

So each colinear block is separated by an equal sign. And each genome segment starts with `>` so we can grab them like so:

```
grep '^>\|^=' wMelwRi_genomes.xmfa > entries_only.xmfa
```

This gives us this:
```
❯ head entries_only.xmfa
> 1:1-46036 + /Users/shelbilrussell/Google Drive/My Drive/lab/reference_seqs/Wolbachia/genomes/wMel-Drosophila_melanogaster/GCF_000008025.1_ASM802v1_genomic.fna
> 2:1-46537 + /Users/shelbilrussell/Google Drive/My Drive/lab/reference_seqs/Wolbachia/genomes/wRi-Drosophila_simulans/GCF_000022285.1_ASM2228v1_genomic.fna
=
> 1:46077-46591 - /Users/shelbilrussell/Google Drive/My Drive/lab/reference_seqs/Wolbachia/genomes/wMel-Drosophila_melanogaster/GCF_000008025.1_ASM802v1_genomic.fna
> 2:76071-76577 + /Users/shelbilrussell/Google Drive/My Drive/lab/reference_seqs/Wolbachia/genomes/wRi-Drosophila_simulans/GCF_000022285.1_ASM2228v1_genomic.fna
=
> 1:49637-50061 - /Users/shelbilrussell/Google Drive/My Drive/lab/reference_seqs/Wolbachia/genomes/wMel-Drosophila_melanogaster/GCF_000008025.1_ASM802v1_genomic.fna
> 2:47889-48312 + /Users/shelbilrussell/Google Drive/My Drive/lab/reference_seqs/Wolbachia/genomes/wRi-Drosophila_simulans/GCF_000022285.1_ASM2228v1_genomic.fna
=
> 1:51489-74728 + /Users/shelbilrussell/Google Drive/My Drive/lab/reference_seqs/Wolbachia/genomes/wMel-Drosophila_melanogaster/GCF_000008025.1_ASM802v1_genomic.fna
```

We can then use this information to create recombinants. For example, the first entry tells us that in wMel `1:46036` is colinear to wRi `1:46537`.  So we can create two recombinant genomes where these are swapped. Because either wmel or wri can be acceptor/donor.
